### PR TITLE
Fix log streaming missing frames

### PR DESCRIPTION
**Pull request by arkadiuss**
Hi,

this MR aims to address the issue reported here https://github.com/hashicorp/nomad/issues/11653 .  I described the problem in [this comment](https://github.com/hashicorp/nomad/issues/11653#issuecomment-996874918). 

This solution adds one more channel to the `streamFile` function that instead of cancelling on the event from channel when eof is reached, it performs one more read until eof which I think should be an expected behavior while streaming logs.

It is a draft for now opened for discussion. I confirmed that it works for the example posted in the issue. It should work since while newFileEvent is generated, the previous file is already full.

 I haven't checked tests yet.

### CODE REVIEW

File: client/fs_endpoint.go [hunk 1]
```diff
@@ -588,11 +590,11 @@ func (f *FileSystem) logsImpl(ctx context.Context, follow, plain bool, offset in
 			close(eofCancelCh)
 			exitAfter = true
 		} else {
-			eofCancelCh = blockUntilNextLog(ctx, fs, logPath, task, logType, idx+1)
+			newFileChannelCh = blockUntilNextLog(ctx, fs, logPath, task, logType, idx+1)
```
Comment (tgross) [inline client/fs_endpoint.go:593]: We've swapped out the `eofCancelCh` for the `newFileChannelCh` here, but now as far as I can tell there's nothing that'll ever send on `eofCancelCh`?

File: client/fs_endpoint.go [hunk 2]
```diff
@@ -637,10 +639,10 @@ func (f *FileSystem) logsImpl(ctx context.Context, follow, plain bool, offset in
 
 // streamFile is the internal method to stream the content of a file. If limit
 // is greater than zero, the stream will end once that many bytes have been
-// read. eofCancelCh is used to cancel the stream if triggered while at EOF. If
+// read. eofCancelOnNextEofCh, if triggered while at EOF, is used to trigger one more read and cancel the stream on reaching next EOF. If
 // the connection is broken an EPIPE error is returned
 func (f *FileSystem) streamFile(ctx context.Context, offset int64, path string, limit int64,
-	fs allocdir.AllocDirFS, framer *sframer.StreamFramer, eofCancelCh chan error) error {
+	fs allocdir.AllocDirFS, framer *sframer.StreamFramer, eofCancelCh chan error, eofCancelOnNextEofCh chan error) error {
```
Comment (tgross) [inline client/fs_endpoint.go:645]: Naming the parameter `eofCancelOnNextEofCh` here when the channel that we construct to pass as that parameter is called `newFileChannelCh` makes this kind of hard to follow. Particularly because we have a similarly named channel right next to it and in the caller.

**Review by tgross — CHANGES_REQUESTED** [review]
Hi @arkadiuss and thanks for this PR! I see where you're going with this but I've left some remarks about the previous channel that we got from `blockUntilNextLog` and what we're doing with it.

File: client/fs_endpoint.go (hunk 2 above)
Comment (arkadiuss) [reply client/fs_endpoint.go:646]: Sure. I renamed the variable to match the naming

File: client/fs_endpoint.go (hunk 1 above)
Comment (arkadiuss) [reply client/fs_endpoint.go]: I think it could be sent. When the follow flag is set to false then a closed channel is passed to end the loop when eof is reached

**Comment by arkadiuss** [comment]
I wanted to cover this scenario in tests. It was hard to create deterministic test since we don't know which event will come first, but I modfied the existing test in a way that the old code (without fix) fails it in most of the cases.

File: client/fs_endpoint.go (hunk 1 above)
Comment (tgross) [reply client/fs_endpoint.go]: Ah yeah I see it now. But given that it will _always_ be closed if it's non-nil, does it really need to be a channel at all now? We could pass a bool that sets the `cancelReceived`, so that as soon as we hit a EOF we return without having to set up the watcher. (That would let us further clean up the channel name too, because it becomes `eofCancelCh` again without confusion.)

Comment (arkadiuss) [reply client/fs_endpoint.go]: Sorry for the delay. I agree with you and applied this comment. 

We could even avoid adding a separate flag  by making an assumption that if `eofCancelCh` is closed then we return with additional read and perform one more when on standard event. But current way with a boolean is more explicit and easier to read probably.

File: client/fs_endpoint.go [hunk 3]
```diff
@@ -704,6 +703,11 @@ OUTER:
 			continue
 		}
 
+		// When eof and cancel received then cancel
```
Comment (tgross) [inline client/fs_endpoint.go:706]: I took a crack at expanding this comment for clarity:

```suggestion
		// At this point we can stop without waiting for more changes,
		// because we have EOF and either we're not following at all,
		// or the channel returned by blockUntilNextLog has closed
```

File: client/fs_endpoint.go [hunk 4]
```diff
@@ -752,12 +756,13 @@ OUTER:
...
 				return nil
-			case err, ok := <-eofCancelCh:
+			case _, ok := <-eofCancelCh:
 				if !ok {
 					return nil
 				}
 
-				return err
+				cancelReceived = true
+				continue OUTER
```
Comment (tgross) [inline client/fs_endpoint.go:765]:
```suggestion
				// try to read one more frame to avoid dropped entries
				// during log rotation
				cancelReceived = true
				continue OUTER
```

File: client/fs_endpoint.go [hunk 5]
```diff
@@ -637,10 +633,10 @@ func (f *FileSystem) logsImpl(ctx context.Context, follow, plain bool, offset in
 
 // streamFile is the internal method to stream the content of a file. If limit
 // is greater than zero, the stream will end once that many bytes have been
-// read. eofCancelCh is used to cancel the stream if triggered while at EOF. If
-// the connection is broken an EPIPE error is returned
+// read. eofCancelCh, if triggered while at EOF, is used to trigger one more read and cancel the stream on reaching next EOF.
+// If the connection is broken an EPIPE error is returned
```
Comment (tgross) [inline client/fs_endpoint.go:637]:
```suggestion
// read. If eofCancelCh is triggered while at EOF, read one more frame and
// cancel the stream on the next EOF. If the connection is broken an EPIPE
// error is returned.
```

File: client/fs_endpoint.go [hunk 6]
```diff
@@ -752,12 +756,13 @@ OUTER:
 				return nil
 			case <-ctx.Done():
 				return nil
-			case err, ok := <-eofCancelCh:
+			case _, ok := <-eofCancelCh:
 				if !ok {
 					return nil
 				}
 
-				return err
```
Comment (tgross) [inline client/fs_endpoint.go:760]: I took a pass through `blockUntilNextLog` and aren't we swallowing non-EOF errors if we do this?

**Review by tgross — CHANGES_REQUESTED** [review]

File: client/fs_endpoint.go (hunk 1 above)
Comment (tgross) [reply client/fs_endpoint.go]: No worries on the delay... we've all been off for the end-of-year anyways 😀

File: client/fs_endpoint.go (hunk 3 above)
Comment (arkadiuss) [reply client/fs_endpoint.go:707]: In the last line I would say that "eof was received and last read was executed", because it doesn't include `blockUntilNextLog` (it's an internal name from a different function) and the information that it's closed is incorrect (it only emitted)

File: client/fs_endpoint.go (hunk 6 above)
Comment (arkadiuss) [reply client/fs_endpoint.go:760]: We do. I will fix it

Comment (arkadiuss) [reply client/fs_endpoint.go:760]: I pushed a change which return err if it's not nil

**Comment by tgross** [comment]
@arkadiuss if you'd like to add a changelog bugfix entry that'd be appreciated. Create a file at `.changelog/11721.txt` that looks something like [this](https://github.com/hashicorp/nomad/blob/main/.changelog/11741.txt). With language something like:

> client: Fixed a bug where the allocation log streaming API was missing log lines that spanned log file rotation

File: client/fs_endpoint.go (hunk 3 above)
Comment (arkadiuss) [reply client/fs_endpoint.go]: Updated

**Comment by arkadiuss** [comment]
> @arkadiuss if you'd like to add a changelog bugfix entry that'd be appreciated. Create a file at `.changelog/11721.txt` that looks something like [this](https://github.com/hashicorp/nomad/blob/main/.changelog/11741.txt). With language something like:
> 
> > client: Fixed a bug where the allocation log streaming API was missing log lines that spanned log file rotation

Thanks, it's added

**Review by tgross — APPROVED** [review]
LGTM! Thanks for the PR @arkadiuss and thanks for sticking with it through a tricky review! This will ship in the next scheduled release of Nomad.

**Comment by arkadiuss** [comment]
> LGTM! Thanks for the PR @arkadiuss and thanks for sticking with it through a tricky review! This will ship in the next scheduled release of Nomad.

Awesome! It was great experience to work with you on it. Thank you for this, in-depth questions and quick responses!
